### PR TITLE
[node] Secure deposit

### PR DESCRIPTION
**Pull request by cf19drofxots**
### Description

This adds secure depositing to a channel, instead of the current naive implementation of simply sending some funds into a channel.

Note this also adds a method to the Node to retrieve a channel's free balance. Useful in checking the owners' balances.

- [x] Deploy preview is functional

### CODE REVIEW

File: packages/node/src/methods/state-channel/deposit/operation.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    multisig: params.multisigAddress,
+    threshold: Zero
+  };
+  const stateChannel = await store.getStateChannel(params.multisigAddress);
+  const stateChannelsMap = await instructionExecutor.runInstallProtocol(
+    new Map<string, StateChannel>([
+      // TODO: (architectural decision) Should this use `getAllChannels` or
+      //       is this good enough? InstallProtocol only operates on a single
+      //       channel, anyway. PR #532 might make this question obsolete.
+      [stateChannel.multisigAddress, stateChannel]
```
Comment (snario) [inline packages/node/src/methods/state-channel/deposit/operation.ts]: yeah I'd just do `getAllChannel`, doesn't hurt

File: packages/node/src/methods/state-channel/deposit/operation.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      respondingAddress: peerAddress,
+      multisigAddress: stateChannel.multisigAddress,
+      aliceBalanceDecrement: Zero,
+      bobBalanceDecrement: Zero,
+      signingKeys: stateChannel.getNextSigningKeys(),
+      terms: {
+        // TODO: generalize
+        assetType: AssetType.ETH,
+        // TODO: is this limit right?
+        limit: params.amount,
```
Comment (snario) [inline packages/node/src/methods/state-channel/deposit/operation.ts]: max uint

File: packages/node/src/methods/state-channel/deposit/operation.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+  const [peerAddress] = await getPeersAddressFromChannel(
+    publicIdentifier,
+    store,
+    params.multisigAddress
+  );
+  const initialState: ETHBalanceRefundAppState = {
+    recipient: publicIdentifier,
+    multisig: params.multisigAddress,
+    threshold: Zero
```
Comment (snario) [inline packages/node/src/methods/state-channel/deposit/operation.ts]: threshold should be `await multisig.getBalance()`

File: packages/node/src/methods/state-channel/deposit/operation.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    store
+  } = requestHandler;
+
+  const [peerAddress] = await getPeersAddressFromChannel(
+    publicIdentifier,
+    store,
+    params.multisigAddress
+  );
+  const initialState: ETHBalanceRefundAppState = {
+    recipient: publicIdentifier,
```
Comment (snario) [inline packages/node/src/methods/state-channel/deposit/operation.ts]: recipient should be the free balance eth account

File: packages/node/src/methods/state-channel/deposit/operation.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    multisig: params.multisigAddress,
```
Comment (snario) [inline packages/node/src/methods/state-channel/deposit/operation.ts]: dont use the params, use the official value from the DB (`stateChannel.multisigAddress`)

File: packages/node/src/methods/state-channel/deposit/operation.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        token: AddressZero
+      },
+      appInterface: {
+        addr: networkContext.ETHBalanceRefund,
+        stateEncoding:
+          "tuple(address recipient, address multisig,  uint256 threshold)",
+        actionEncoding: undefined
+      },
+      // TODO: what should the default timeout be here?
+      defaultTimeout: One.toNumber()
```
Comment (snario) [inline packages/node/src/methods/state-channel/deposit/operation.ts]: do 7 days for now

File: packages/node/src/methods/state-channel/deposit/operation.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    }
+  );
+
+  await store.saveStateChannel(stateChannelsMap.get(params.multisigAddress)!);
```
Comment (snario) [inline packages/node/src/methods/state-channel/deposit/operation.ts]: avoid `!` when possible an do an error check

File: packages/node/src/methods/state-channel/deposit/operation.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    to: params.multisigAddress,
+    value: params.amount
+  };
+
+  const depositPromise = requestHandler.wallet.sendTransaction({
+    ...tx,
+    gasPrice: await requestHandler.provider.estimateGas(tx)
+  });
+  requestHandler.outgoing.emit(NODE_EVENTS.DEPOSIT_STARTED);
+  depositPromise.then(async () => {
```
Comment (snario) [inline packages/node/src/methods/state-channel/deposit/operation.ts]: just use await?

File: packages/node/src/methods/state-channel/deposit/operation.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+  const stateChannelsMap = await instructionExecutor.runUninstallProtocol(
+    new Map(Object.entries(await store.getAllChannels())),
+    {
+      initiatingAddress: publicIdentifier,
+      respondingAddress: peerAddress,
+      multisigAddress: stateChannel.multisigAddress,
+      appIdentityHash: stateChannel.getBalanceRefund(
+        requestHandler.networkContext
+      ).identityHash,
+      aliceBalanceIncrement: params.amount,
```
Comment (snario) [inline packages/node/src/methods/state-channel/deposit/operation.ts]: be certain it is alice and not bob; its the sorted 0th child keys

File: packages/node/src/methods/state-channel/deposit/operation.ts (hunk 7 above)
Comment (cf19drofxots) [reply packages/node/src/methods/state-channel/deposit/operation.ts]: This is mostly copy pasted from the already-existing install code.

File: packages/node/src/methods/state-channel/deposit/operation.ts (hunk 8 above)
Comment (cf19drofxots) [reply packages/node/src/methods/state-channel/deposit/operation.ts]: Doing an immediate await wouldn't allow for emitting the event.

**Comment by cf19drofxots** [comment]
The failure is not relevant to this PR and seems to have been introduced when master was merged: 9fb8cddff282a5d949f5d76862fb424433dbb912

**Pull request merged**